packages/rulesets/src/spectral/functions/validate-patch-body-param-properties.ts

```diff
@@ -46,19 +46,39 @@ export const validatePatchBodyParamProperties = createRulesetFunction<unknown, O
         _opts.should.forEach((p: string) => {
           if (!getProperties(bodyParameter)?.[p] && getProperties(responseSchema)?.[p]) {
             errors.push({
-              message: `The patch operation body parameter schema should contains property '${p}'.`,
+              message: `The patch operation body parameter schema should contain property '${p}'.`,
               path: [...path, "parameters", index],
             })
           }
         })
       }
       if (_opts.shouldNot) {
         _opts.shouldNot.forEach((p: string) => {
-          if (getProperties(bodyParameter)?.[p]) {
-            errors.push({
-              message: `The patch operation body parameter schema should not contains property ${p}.`,
-              path: [...path, "parameters", index],
-            })
+          const property = getProperties(bodyParameter)?.[p];
+          if (property) {
+            let isPropertyBoolean = false;
+            let isPropertyImmutable = false;
+
+            if (property["readOnly"] && property["readOnly"] === true)
+            {
+              isPropertyBoolean = true;
+            }
+
+            if (property["x-ms-mutability"] && Array.isArray(property["x-ms-mutability"])) {
+                const schemeArray = property["x-ms-mutability"];
+                if (!schemeArray.includes("update"))
+                {
+                    isPropertyImmutable = true;
+                }
+            }
+
+            if (!isPropertyBoolean && !isPropertyImmutable)
+            {
+              errors.push({
+                message: `The patch operation body parameter schema should not contain property ${p}.`,
```

Comment:
Can we provide more information here? Like:

`message: 'The patch operation body parameter schema should not contain property ${p}, because the property is not readable nor immutable.'`,

Also, I am confused: the doc for the rule says that the property should not be included in parameters if `it is read-only or it is immutable`, but this code makes me think the property should not be included in parameters `unless it is read-only or it is immutable`. Which one is it?

Maybe instead `if (!isPropertyBoolean && !isPropertyImmutable)`
we should have `if (isPropertyReadOnly || isPropertyImmutable)` ? 

Then it would make sense to me: if a property is read-only or immutable, we notify the user it should not be in the params.

Reply:
OK, I think a little bit too many nested negations threw me off ;) In particular the `_opts.shouldNot` is a bit misleading. Because the property **can** be included even if it **should not**. But if this is the case, then it has to be `readOnly || immutable`.

Btw. maybe we could clarify the intent of `if (!(isPropertyReadonly || isPropertyImmutable))` by extracting a method, like `if (isPropertyWritable())` ? One negation less :)